### DISCUSSION TITLE
Discuss about Strategies and Segments implemetations

### DISCUSSION THREAD

**Discussion by kpicaza** [pheature-flags/pheature-flags#156] [Ideas]
**Epic:** https://github.com/pheature-flags/pheature-flags/issues/2 Toggle Model
**Package:** https://github.com/pheature-flags/toggle-model 

**Description:**

At this moment, we have provisional implementations for both object types.

We need to discuss:

* How we can left end users adding custom strategies without breaking our strictness policies.
* What types of strategies should be shipped as core models.
* What types of segments should be shipped as core models.

[UPDATE] Let check Martin Fowlers article on subject https://martinfowler.com/articles/feature-toggles.html

**Comment by kpicaza** [comment]
**Proposal**

**Toggle Strategies**

* Create ChainToggleStrategyFactory class that allows loading differents Strategy type factories, configured by the ToggleConfig object
* Move ToggleConfig class from [PRS-11 factories package](https://github.com/pheature-flags/toggle-psr11-factories) to [Toggle Model](https://github.com/pheature-flags/toggle-model) to allow loading user created ToggleStrategyFactory instances.
* Create a ToggleStrategyFactory that allows creating "core" Strategy type instances
* Check EnableByMatchingSegment
* Update EnableByMatchingIdentity
* Create EnableByPercentage
* Create ChainToggleStrategy: Chain strategy adds the ability to mix different strategies with an "AND" 
* Document in detail on the architecture section of the docs
* Document How to create and share in community Custom Strategies also in the architecture section of the docs

**Segments**

* Create ChainSegmentFactory class that allows loading differents Segment type factories, configured by the ToggleConfig object
* Create a SegmentFactory that allows creating "core" Segment type instances
* Rename current Segment to ExactMatchingSegment
* Create CollectionMatchingSegment
* Create ChainSegment class that adds the ability to mix different Segments with an "AND" 
* Document in detail on the architecture section of the docs
* Document How to create and share in community Custom Segments also in the architecture section of the docs

**Community**

* Create PSR-11 container compatible factories for ChainToggleStrategyFactory and ChainSegmentFactory

![image](https://user-images.githubusercontent.com/1093654/114610071-cdb73080-9c9f-11eb-80d3-a58a75936a91.png)

**Comment by xserrat** [comment]
Hi @kpicaza !

Nice proposal 👏🏽  Could you make the [toggle-psr11-factories](https://github.com/pheature-flags/toggle-psr11-factories) repo accessible to take a look on these factories? 

Regarding the proposal I have a doubt about the factories:
* The `ChainToggleStrategyFactory` should create a `ChainToggleStrategy` based on the `ToggleConfig` that should contain the strategies to load, right? Are these strategies passed to the `ChainToggleStrategy` as factories instead of `Strategy` instances to reduce the coupling with the container and use lazy load to build each `Strategy` instance? Or there's another reason?

Also, to make easier the combination of strategies for the users who need a complex strategy, we could add specifications for the `OR` and `AND` operations like these examples:
* [OR strategy](https://github.com/DesignPatternsPHP/DesignPatternsPHP/blob/main/Behavioral/Specification/OrSpecification.php)
* [AND strategy](https://github.com/DesignPatternsPHP/DesignPatternsPHP/blob/main/Behavioral/Specification/AndSpecification.php)

so then they'll be able to split parts of the strategy in multiple classes. What do you think?


Thanks and good job 🚀 !!!

**Comment by kpicaza** [comment]
Hi @xserrat :wave: 

> Could you make the toggle-psr11-factories repo accessible to take a look on these factories?


I'm sorry, is a typo, the repo is public and accessible here https://github.com/pheature-flags/toggle-crud-psr11-factories

Happy to see you here ;-D

**Comment by kpicaza** [comment]
> Are these strategies passed to the ChainToggleStrategy as factories instead of Strategy instances to reduce the coupling with the container and use lazy load to build each Strategy instance? Or there's another reason?

It's right @xserrat, also the idea is to allow the end-user(the developer) to define its own strategy types(red boxes) that can be loaded from the container maintaining the core strict type coherence. The same logic applies to the segments. We can search for a less complex approach here,

> Also, to make easier the combination of strategies for the users who need a complex strategy, we could add specifications for the OR and AND operations like these examples:

PR or pair programming session welcome;-D. We put an "AND" at this level because the "OR" is applied in the Toggle by design. It is a good point. 

Also, take a look at the Flow picture in the docs https://pheatureflags.io/#/architecture/?id=flow
![](https://pheatureflags.io/assets/toggle-flow-diagram.png)

Thank you very much.
